cmd/faucet/faucet.go

```diff
@@ -178,288 +134,164 @@ func validAddress(addr string) error {
-				log.Printf("error sending batch: %s", err)
-			}
-		}
+func InitFaucet(logger zerolog.Logger) (Faucet, error) {
+	var err error
+	cfg, err := config.LoadConfig()
+	if err != nil {
+		panic(err)
 	}
-}
 
-func (c *Client) sendBatchIfNeededLock(ctx context.Context) error {
-	c.batchmu.Lock()
-	defer c.batchmu.Unlock()
-	return c.sendBatchIfNeeded(ctx)
-}
-
-func (c *Client) sendBatchIfNeeded(ctx context.Context) error {
-	if len(c.batch) == 0 {
-		return nil
+	totalSupply, err := strconv.Atoi(cfg.DailyLimit)
+	if err != nil {
+		return Faucet{}, err
+	}
+	amount, err := strconv.Atoi(cfg.Amount)
+	if err != nil {
+		return Faucet{}, err
 	}
 
-	err := c.sendBatch(ctx)
-	c.batch = nil
-	return err
-}
-
-func (c *Client) sendBatch(ctx context.Context) error {
-	log.Printf("sending %s to %v", c.cfg.SendDenom, DestAddresses(c.batch))
-	send := "send"
-	if len(c.batch) > 1 {
-		send = "multi-send"
+	f := Faucet{
+		config:          cfg,
+		Mutex:           &sync.Mutex{},
+		Batch:           []string{},
+		log:             logger,
+		TokensAvailable: float64(totalSupply) / uwardConversion,
+		DailySupply:     float64(totalSupply) / uwardConversion,
+		Amount:          float64(amount),
 	}
-	cmd := strings.Join([]string{
-		c.cfg.CliName,
-		"tx",
-		"bank",
-		send,
-		c.cfg.AccountName,
-		strings.Join(DestAddresses(c.batch), " "),
-		c.cfg.SendDenom,
-		"--yes",
-		"--keyring-backend",
-		c.cfg.KeyringBackend,
-		"--chain-id",
-		c.cfg.ChainID,
-		"--node",
-		c.cfg.Node,
-		"--gas-prices",
-		c.cfg.Fees,
-		"-o",
-		"json",
-	}, " ")
 
-	out, err := e(ctx, cmd, false)
+	f.Cooldown, err = time.ParseDuration(f.config.CoolDown)
 	if err != nil {
-		for _, dest := range c.batch {
-			dest.Reset()
-		}
-		return err
+		return Faucet{}, err
 	}
 
-	var result struct {
-		Code   int    `json:"code"`
-		TxHash string `json:"txhash"`
-	}
-	if err := json.Unmarshal(out.Stdout, &result); err != nil {
-		for _, dest := range c.batch {
-			dest.Reset()
-		}
-		return fmt.Errorf("error unmarshalling tx result: %w", err)
-	}
-	if result.Code != 0 {
-		for _, dest := range c.batch {
-			dest.Reset()
-		}
-		return fmt.Errorf("tx failed with code %d", result.Code)
+	f.BatchInterval, err = time.ParseDuration(f.config.BatchInterval)
+	if err != nil {
+		return Faucet{}, err
 	}
 
-	err = c.waitTx(ctx, result.TxHash)
-	if err != nil {
-		for _, dest := range c.batch {
-			dest.Reset()
+	if f.config.Mnemonic != "" {
+		if err = f.setupNewAccount(); err != nil {
+			return Faucet{}, err
 		}
-		return fmt.Errorf("error waiting for tx: %w", err)
 	}
 
-	return nil
+	return f, nil
 }
 
-func (c *Client) waitTx(ctx context.Context, txHash string) error {
-	cmd := strings.Join([]string{
-		c.cfg.CliName,
-		"q",
-		"tx",
-		txHash,
-		"--node",
-		c.cfg.Node,
-		"-o",
-		"json",
-	}, " ")
+func MutexLocked(m *sync.Mutex) bool {
+	state := reflect.ValueOf(m).Elem().FieldByName("state")
+	return state.Int()&mutexLocked == mutexLocked
```

Comment:
i think that generally speaking this is a very bad practice, as this method itself is not thread safe 😁 

MutexLocked could return `false`. But immediately after the mutex might get locked.

If you accept this, it's okay. My comment is to acknowledge that this isn't really a protection and two f.Send() might still occur simultaneously.

Reply:
Got it, removed the lock check